python/src/iceberg/transforms.py

```diff
@@ -301,6 +301,86 @@ def _(self, value_type: IcebergType, value: int) -> str:
         return datetime.to_human_timestamptz(value)
 
 
+class TruncateTransform(Transform[S, S]):
+    """A transform for truncating a value to a specified width.
+    Args:
+      source_type (Type): An Iceberg Type of IntegerType, LongType, StringType, BinaryType or DecimalType
+      width (int): The truncate width
+    Raises:
+      ValueError: If a type is provided that is incompatible with a Truncate transform
+    """
+
+    def __init__(self, source_type: IcebergType, width: int):
+        super().__init__(
+            f"truncate[{width}]",
+            f"transforms.truncate(source_type={repr(source_type)}, width={width})",
+        )
+        self._type = source_type
+        self._width = width
+
+    @property
+    def width(self):
+        return self._width
+
+    def apply(self, value: Optional[S]) -> Optional[S]:
+        return self._truncate_value(value) if value is not None else None
+
+    @singledispatchmethod
+    def _truncate_value(self, value: S) -> S:
+        raise ValueError(f"Cannot truncate value: {value}")
+
+    @_truncate_value.register(int)
+    def _(self, value):
```

Comment:
```suggestion
    def _(self, value: int) -> int:
```

Reply:
@Fokko I tried it before but mypy threw an error
```
an python/src/iceberg/transforms.py:332: error: Argument 1 has incompatible type "Callable[[TruncateTransform[S], int], int]"; expected "Callable[..., S]"
```
Wondering if there is a way to solve it?

Reply:
updated it using `singledispatch` here and then use Any for value.